src/engine/sys/sdl_glimp.cpp

```diff
@@ -2154,6 +2156,217 @@ static void reportHardwareType( bool force )
+		{
+			glConfig2.driverVendor = glDriverVendor_t::MESA;
+			glConfig2.hardwareVendor = p.second;
+			return;
+		}
+	}
+	
+	// Those vendor strings are assumed to be unambiguous about being from Mesa drivers.
+	static const std::string mesaVendors[] = {
+		"Mesa",
+		"Mesa Project",
+		"Mesa/X.org",
+		"X.Org",
+		// Zink.
+		"Collabora Ltd",
+		// virgl.
+		"Red Hat",
+		// llvmpipe, softpipe, SVGA3D.
+		"VMware, Inc.",
+	};
+
+	for ( auto& s : mesaVendors )
+	{
+		if ( !strcmp( glConfig.vendor_string, s.c_str() ) )
+		{
+			glConfig2.driverVendor = glDriverVendor_t::MESA;
+		}
+	}
+
+	/* This substring in a version string is assumed to be unambiguous about being from Mesa
+	drivers. Mesa GL_VENDOR and GL_RENDERER strings are very fluid, but it is believed that
+	GL_VERSION always contains the " Mesa " substring and that substring always means Mesa. */
+	if ( !Util::ordinal(glConfig2.driverVendor) && strstr( glConfig.version_string, " Mesa " ) )
+	{
+		glConfig2.driverVendor = glDriverVendor_t::MESA;
+	}
+
+	if ( glConfig2.driverVendor == glDriverVendor_t::MESA )
+	{
+		// This vendor string for a Mesa driver is assumed to be unambiguous about the hardware.
+		if ( !strcmp( glConfig.vendor_string, "Intel" ) )
+		{
+			glConfig2.hardwareVendor = glHardwareVendor_t::INTEL;
+			return;
+		}
+
+		/* Those substrings at the beginning a renderer string are assumed to be unambiguous about
+		the hardware or underlying technology. */
+		if ( !Util::ordinal(glConfig2.hardwareVendor) )
+		{
+			static const std::pair<std::string, glHardwareVendor_t> rendererStartStringHardware[] = {
+				{ "ATI ", glHardwareVendor_t::ATI },
+				{ "AMD ", glHardwareVendor_t::ATI },
+				{ "NV", glHardwareVendor_t::NVIDIA },
+				{ "i915 ", glHardwareVendor_t::INTEL },
+				// OpenGL over Vulkan translation.
+				{ "zink ", glHardwareVendor_t::TRANSLATION },
+				// Virtualization.
+				{ "virgl", glHardwareVendor_t::TRANSLATION },
+			};
+
+			for ( auto& p : rendererStartStringHardware )
+			{
+				if ( !strncmp( glConfig.renderer_string, p.first.c_str(), p.first.size() ) )
+				{
+					glConfig2.hardwareVendor = p.second;
+					return;
+				}
+			}
+		}
+
+		/* Those substrings within a renderer string are assumed to be unambiguous about
+		the underlying technology. */
+		if ( !Util::ordinal(glConfig2.hardwareVendor) )
+		{
+			static const std::pair<std::string, glHardwareVendor_t> rendererSubStringHardware[] = {
+				{ "llvmpipe", glHardwareVendor_t::SOFTWARE },
+				{ "softpipe", glHardwareVendor_t::SOFTWARE },
+				// Virtualization.
+				{ "SVGA3D", glHardwareVendor_t::TRANSLATION },
+			};
+
+			for ( auto& p : rendererSubStringHardware )
+			{
+				if ( strstr( glConfig.renderer_string, p.first.c_str() ) )
+				{
+					glConfig2.hardwareVendor = p.second;
+					return;
+				}
+			}
+		}
+	}
+
+	/* As both proprietary Intel driver on Windows and Mesa may report "Intel",
+	we rely on the fact Mesa is already detected to know it is the proprietary
```

Comment:
It seems the function can reach this point with Mesa detected but no hardware vendor.

Reply:
If it's about the risk of a Mesa detected driver to match the following Intel test, I can test for `!Mesa` but if `Mesa and Intel`, the code is supposed to already have returned before.

If it's simply about the risk of reaching that point without hardware being detected, it's fine. The code doesn't guarantee to detect something or to exhaustively know about everything, it only tries to guarantee that if something is detected, we can trust what it is said to be detected.